### PR TITLE
Implement support for .debug_pubnames/.debug_pubtypes

### PR DESCRIPTION
**Pull request by khuey**
I wrote this between 11 pm - 2 am last night, so while I thought the generics are clever, I might just be going off the rails here :P

### CODE REVIEW

File: src/lookup.rs [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+    remaining_input: EndianBuf<'input, Endian>,
+}
+
+impl<'input, Endian, Parser> LookupEntryIter<'input, Endian, Parser>
+    where Endian: Endianity,
+          Parser: LookupParser<'input, Endian>
+{
+    /// Advance the iterator and return the next entry.
+    ///
+    /// Returns the newly parsed arange as `Ok(Some(arange))`. Returns
```
Comment (fitzgen) [inline src/lookup.rs]: s/arange/Parser::Entry/

File: src/lookup.rs [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+                            self.next_entry()
+                        }
+                        Some(entry) => Ok(Some(entry)),
+                    }
+                }
+                Err(e) => {
+                    self.current_set = self.current_set.range_to(..0);
+                    self.current_header = None;
+                    // Should we blow away all other sets too? Maybe not ...
+                    self.remaining_input = self.remaining_input.range_to(..0);
```
Comment (fitzgen) [inline src/lookup.rs]: Probably. I'd expect getting an error to make this type unusable from then on.

File: src/lookup.rs [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+            (Ordering::Equal, Ordering::Equal, Ordering::Less) => Ordering::Less,
+            (Ordering::Greater, _, _) |
+            (Ordering::Equal, Ordering::Greater, _) |
+            (Ordering::Equal, Ordering::Equal, Ordering::Greater) => Ordering::Greater,
+        }
+    }
+}
+
+pub struct ArangeParser<'input, Endian: 'input>
+    where Endian: Endianity
```
Comment (fitzgen) [inline src/lookup.rs]: Nit: move the `'input` bound into the where clause like this:

```rust
pub struct ArangeParser<'input, Endian>
    where Endian: 'input + Endianity
{
    ...
}
```

File: src/lookup.rs [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+
+    fn parse_offset(input: EndianBuf<Endian>,
+                    format: Format)
+                    -> ParseResult<(EndianBuf<Endian>, Self::Offset)>;
+
+    fn format_from(header: &Self::Header) -> Format;
+}
+
+pub struct PubStuffParser<'input, Endian: 'input, Switch: 'input>
+    where Endian: Endianity,
```
Comment (fitzgen) [inline src/lookup.rs]: ditto regarding `'input` lifetime bound

File: src/lookup.rs [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+        };
+        let dividing_line: usize = try!(set_length.checked_sub(header_length)
+            .ok_or(Error::BadLength)) as usize;
+
+        Ok((rest.range_from(dividing_line..),
+            rest.range_to(..dividing_line),
+            Switch::new_header(format, set_length, version, info_offset, info_length)))
+    }
+
+    /// Parse a single pubname. Return `None` for the null pubname, `Some` for an actual pubname.
```
Comment (fitzgen) [inline src/lookup.rs]: s/pubname/pubfoo/

File: src/parser.rs [hunk 6]
```diff
@@ -231,6 +231,20 @@ pub fn parse_uN_as_u64<Endian>(size: u8,
     }
 }
 
+/// Parse either a `u32` or a `u64`, depending on the format, and return it as a `u64`
+#[doc(hidden)]
+#[inline]
+pub fn parse_format_size_as_u64<Endian>(input: EndianBuf<Endian>,
```
Comment (fitzgen) [inline src/parser.rs]: This can be replaced with `parse_word` thanks to the other PR that just merged.

**Comment by fitzgen** [comment]
LGTM!

This needs to be rebased over recent PRs.

I think it would be nice to separate the generic lookup parser module from the concrete parsers that use it. In other words, still have `aranges.rs` that is pretty thin and mostly just calls into the generic parser infrastructure. Same for `pubnames.rs` and `pubtypes.rs`.

**Comment by khuey** [comment]
Comments addressed!

**Comment by fitzgen** [comment]
Awesome! Thanks!

**Pull request merged**